akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -842,61 +842,80 @@ final case class Sliding[T](val n: Int, val step: Int) extends GraphStage[FlowSh
-    else ctx.push(buffer.dequeue())
-  }
-
-  override def onUpstreamFinish(ctx: DetachedContext[T]): TerminationDirective =
-    if (buffer.isEmpty) ctx.finish()
-    else ctx.absorbTermination()
-
-  val enqueueAction: (DetachedContext[T], T) ⇒ UpstreamDirective =
-    overflowStrategy match {
-      case DropHead ⇒ (ctx, elem) ⇒
-        if (buffer.isFull) buffer.dropHead()
-        buffer.enqueue(elem)
-        ctx.pull()
-      case DropTail ⇒ (ctx, elem) ⇒
-        if (buffer.isFull) buffer.dropTail()
-        buffer.enqueue(elem)
-        ctx.pull()
-      case DropBuffer ⇒ (ctx, elem) ⇒
-        if (buffer.isFull) buffer.clear()
-        buffer.enqueue(elem)
-        ctx.pull()
-      case DropNew ⇒ (ctx, elem) ⇒
-        if (!buffer.isFull) buffer.enqueue(elem)
-        ctx.pull()
-      case Backpressure ⇒ (ctx, elem) ⇒
-        buffer.enqueue(elem)
-        if (buffer.isFull) ctx.holdUpstream()
-        else ctx.pull()
-      case Fail ⇒ (ctx, elem) ⇒
-        if (buffer.isFull) ctx.fail(new BufferOverflowException(s"Buffer overflow (max capacity was: $size)!"))
-        else {
+    val enqueueAction: T ⇒ Unit =
+      overflowStrategy match {
+        case DropHead ⇒ elem ⇒
+          if (buffer.isFull) buffer.dropHead()
+          buffer.enqueue(elem)
+          pull(in)
+        case DropTail ⇒ elem ⇒
+          if (buffer.isFull) buffer.dropTail()
+          buffer.enqueue(elem)
+          pull(in)
+        case DropBuffer ⇒ elem ⇒
+          if (buffer.isFull) buffer.clear()
+          buffer.enqueue(elem)
+          pull(in)
+        case DropNew ⇒ elem ⇒
+          if (!buffer.isFull) buffer.enqueue(elem)
+          pull(in)
+        case Backpressure ⇒ elem ⇒
           buffer.enqueue(elem)
-          ctx.pull()
+          if (!buffer.isFull) pull(in)
+        case Fail ⇒ elem ⇒
+          if (buffer.isFull) failStage(new BufferOverflowException(s"Buffer overflow (max capacity was: $size)!"))
+          else {
+            buffer.enqueue(elem)
+            pull(in)
+          }
+      }
+
+    override def preStart(): Unit = {
+      buffer = BufferImpl(size, materializer)
+      pull(in)
+    }
+
+    override def onPush(): Unit = {
+      val elem = grab(in)
+      if (isAvailable(out)) {
+        if (buffer.isEmpty) {
+          push(out, elem)
+          pull(in)
+        } else {
+          push(out, buffer.dequeue())
+          enqueueAction(elem)
         }
+      } else {
+        enqueueAction(elem)
+      }
     }
 
+    override def onPull(): Unit = {
+      if (isClosed(in)) {
+        if (buffer.nonEmpty) push(out, buffer.dequeue())
+        if (buffer.isEmpty) completeStage()
+      } else if (!hasBeenPulled(in)) {
+        if (buffer.nonEmpty) push(out, buffer.dequeue())
+        pull(in)
+      } else if (buffer.nonEmpty) {
+        push(out, buffer.dequeue())
+      }
+    }
+
+    override def onUpstreamFinish(): Unit = {
+      if (buffer.isEmpty) completeStage()
+      else if (isAvailable(out)) onPull()
```

Comment:
how can this happen? If the buffer is not empty, then `isAvailable(out)` cannot be true unless we were pulled and forgot to push a dequeued element.